src/librbd/io/ImageRequestWQ.cc

```diff
@@ -629,7 +629,9 @@ void *ImageRequestWQ<I>::_void_dequeue() {
     m_image_ctx.owner_lock.get_read();
     if (m_image_ctx.exclusive_lock != nullptr) {
       ldout(cct, 5) << "exclusive lock required: delaying IO " << item << dendl;
-      if (!m_image_ctx.get_exclusive_lock_policy()->may_auto_request_lock()) {
+      auto auto_request = m_image_ctx.get_exclusive_lock_policy()->may_auto_request_lock();
+      auto is_lock_owner = m_image_ctx.exclusive_lock->is_lock_owner();
```

Comment:
The test on line 627 should have already tested whether or not the lock was required (i.e. not lock owner).

Reply:
.... but that's the expected result. When you set ```rbd_auto_exclusive_lock_until_manual_request``` to false, you are saying that you want to manually acquire the exclusive lock.

Reply:
... since you bring up the Python environment, this is the specific API method you would be required to invoke prior to starting IO [1].

[1] https://github.com/ceph/ceph/blob/v12.2.3/src/pybind/rbd/rbd.pyx#L285

Reply:
rbd_auto_exclusive_lock_until_manual_request to false, you are saying that you want to manually acquire the exclusive lock.
// that's true. but IO should be operated oK after i manually acquired lock,  that's why ``image.lock_acquire(0)``` located there. Indeed, the previous code will return EROFS, even lock is already acquired, which i think is not the expected result.

```image.lock_acquire```(used in test code) calls the same interface with https://github.com/ceph/ceph/blob/v12.2.3/src/pybind/rbd/rbd.pyx#L285.

see
https://github.com/ceph/ceph/blob/2dab17a455c09584f2a85e6b10888337d1ec8949/src/pybind/rbd/rbd.pyx#L2300

do i miss something？

Reply:
sorry, I missed that it was already acquiring the lock. Therefore, the issue is the original issue that I commented about: that if you have already acquired the lock, then ```lock_required``` should be false and you should not have entered that code block. We need to discover why ```lock_required``` is set to true after the lock has already been acquired. The ```set_require_lock``` must be not called somehow.